src/detect-flow-pkts.c

```diff
@@ -23,238 +23,412 @@
+                de_ctx, s, DETECT_FLOW_PKTS, (SigMatchCtx *)df, DETECT_SM_LIST_MATCH) == NULL) {
+        DetectFlowPktsFree(de_ctx, df);
         return -1;
     }
     s->flags |= SIG_FLAG_REQUIRE_PACKET;
 
     return 0;
 }
 
-static void PrefilterPacketFlowPktsToClientMatch(
-        DetectEngineThreadCtx *det_ctx, Packet *p, const void *pectx)
+static int DetectFlowPktsToClientSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    const PrefilterPacketHeaderCtx *ctx = pectx;
-    if (!PrefilterPacketHeaderExtraMatch(ctx, p))
-        return;
+    DetectU32Data *du32 = DetectU32Parse(rawstr);
+    if (du32 == NULL)
+        return -1;
 
-    DetectU32Data du32;
-    du32.mode = ctx->v1.u8[0];
-    du32.arg1 = ctx->v1.u32[1];
-    du32.arg2 = ctx->v1.u32[2];
-    if (DetectFlowPktsToClientMatch(det_ctx, p, NULL, (const SigMatchCtx *)&du32)) {
-        PrefilterAddSids(&det_ctx->pmq, ctx->sigs_array, ctx->sigs_cnt);
-    }
-}
+    DetectFlow *df = SCCalloc(1, sizeof(DetectFlow));
+    if (df == NULL)
+        return -1;
 
-static int PrefilterSetupFlowPktsToClient(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
-{
-    return PrefilterSetupPacketHeader(de_ctx, sgh, DETECT_FLOW_PKTS_TO_CLIENT,
-            SIG_MASK_REQUIRE_FLOW, PrefilterPacketU32Set, PrefilterPacketU32Compare,
-            PrefilterPacketFlowPktsToClientMatch);
-}
+    df->pkt_data = du32;
+    df->dir = DETECT_FLOW_TOCLIENT;
 
-static bool PrefilterFlowPktsToClientIsPrefilterable(const Signature *s)
-{
-    return PrefilterIsPrefilterableById(s, DETECT_FLOW_PKTS_TO_CLIENT);
-}
+    if (SigMatchAppendSMToList(
+                de_ctx, s, DETECT_FLOW_PKTS, (SigMatchCtx *)df, DETECT_SM_LIST_MATCH) == NULL) {
+        DetectFlowPktsFree(de_ctx, df);
+        return -1;
+    }
+    s->flags |= SIG_FLAG_REQUIRE_PACKET;
 
-void DetectFlowPktsToClientRegister(void)
-{
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].name = "flow.pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].desc = "match flow number of packets to client";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].url = "/rules/flow-keywords.html#flow-pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Match = DetectFlowPktsToClientMatch;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Setup = DetectFlowPktsToClientSetup;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Free = DetectFlowPktsToClientFree;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SupportsPrefilter =
-            PrefilterFlowPktsToClientIsPrefilterable;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SetupPrefilter = PrefilterSetupFlowPktsToClient;
+    return 0;
 }
 
-static int DetectFlowPktsToServerMatch(
-        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+static int DetectFlowPktsSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    if (p->flow == NULL) {
-        return 0;
+    DetectFlow *df = NULL;
+    char copy[strlen(rawstr) + 1];
```

Comment:
Could we do this parsing in rust ?

Reply:
ok 😢

Reply:
I think this can be done as a follow up effort.